File: articles/differential-equations-population-physiological-modeling/python/pharmacokinetics.py
"""
One-compartment pharmacokinetic scenarios.

Run:
    python python/pharmacokinetics.py
"""

from pathlib import Path
import math

import pandas as pd

from differential_equations_core import simulate_one_compartment


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "pharmacokinetic_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        trajectory = simulate_one_compartment(row["C0"], row["elimination_rate"], row["dt"], row["t_end"])

        rows.append(
            {
                "scenario": row["scenario"],
                "initial_concentration": trajectory["concentration"].iloc[0],
                "final_concentration": trajectory["concentration"].iloc[-1],
                "elimination_rate": row["elimination_rate"],
                "half_life": math.log(2) / row["elimination_rate"],
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/differential-equations-population-physiological-modeling/r/pharmacokinetics.R
# One-compartment pharmacokinetic model.

simulate_one_compartment <- function(C0, elimination_rate, dt = 0.05, t_end = 48) {
  time <- seq(0, t_end, by = dt)
  C <- numeric(length(time))
  C[1] <- C0

  for (i in 2:length(time)) {
    dC <- -elimination_rate * C[i - 1]
    C[i] <- max(C[i - 1] + dC * dt, 0)
  }

  data.frame(time = time, concentration = C)
}

trajectory <- simulate_one_compartment(20, 0.12)

summary_df <- data.frame(
  initial_concentration = trajectory$concentration[1],
  final_concentration = tail(trajectory$concentration, 1),
  half_life = log(2) / 0.12
)

print(round(summary_df, 5))
